{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Trim:\n",
    "    def __init__(self,forward_fastq,reverse_fastq, barcode_fasta, length, output_dir, output_forward_name, output_reverse_name):\n",
    "        \"\"\"\n",
    "        This function takes in the forward and reverse fastq files, the barcode fasta file, the length\n",
    "        of the barcode, the output directory, and the output file names for the forward and reverse\n",
    "        reads\n",
    "        \n",
    "        :param forward_fastq: The forward fastq file\n",
    "        :param reverse_fastq: The name of the reverse fastq file\n",
    "        :param barcode_fasta: a fasta file containing the barcodes\n",
    "        :param length: the length of the barcode\n",
    "        :param output_dir: The directory where the output files will be saved\n",
    "        :param output_forward_name: the name of the output file for the forward reads\n",
    "        :param output_reverse_name: the name of the output file for the reverse reads\n",
    "        \"\"\"\n",
    "\n",
    "        self.forward_fastq = forward_fastq\n",
    "        self.reverse_fastq = reverse_fastq\n",
    "        self.barcode_fasta = barcode_fasta\n",
    "        self.length = length\n",
    "        self.output_dir = output_dir\n",
    "        self.output_forward_name = output_forward_name\n",
    "        self.output_reverse_name = output_reverse_name\n",
    "\n",
    "    def forward_fastq_reader(self):\n",
    "        \"\"\"\n",
    "        This function takes a fastq file as input and returns a list of sequences\n",
    "        :return: A list of sequences\n",
    "        \"\"\"\n",
    "\n",
    "        n = []\n",
    "        for record in SeqIO.parse(self.forward_fastq , \"fastq\"):\n",
    "            n.append(str(record.seq))\n",
    "        return n\n",
    "\n",
    "    def reverse_fastq_reader(self):\n",
    "        \"\"\"\n",
    "        This function takes a fastq file and returns a list of reverse complemented sequences\n",
    "        :return: A list of the reverse complements of the sequences in the fastq file.\n",
    "        \"\"\"\n",
    "\n",
    "        n = []\n",
    "        for record in SeqIO.parse(self.reverse_fastq , \"fastq\"):\n",
    "            n.append(str(record.seq)[::-1])\n",
    "        return n\n",
    "        \n",
    "    def forward_barcode_reader(self):\n",
    "        \"\"\"\n",
    "        This function takes a fasta file of barcodes and returns a list of the barcodes\n",
    "        :return: A list of strings.\n",
    "        \"\"\"\n",
    "\n",
    "        forward = []\n",
    "        for record in SeqIO.parse(self.barcode_fasta, \"fasta\"):\n",
    "            forward.append(str(record.seq))\n",
    "        return forward\n",
    "    \n",
    "    def reverse_barcode_reader(self):\n",
    "        \"\"\"\n",
    "        This function takes in a list of forward barcodes and returns a list of reverse barcodes\n",
    "        :return: A list of the reverse complements of the barcodes.\n",
    "        \"\"\"\n",
    "     \n",
    "        forward = self.forward_barcode_reader()\n",
    "\n",
    "        reversed = []\n",
    "        for i in forward:\n",
    "            reversed.append(str(Seq(i).reverse_complement())[::-1])\n",
    "        return reversed\n",
    "        \n",
    "    def trimmer(self, complement):\n",
    "        \"\"\"\n",
    "        Given a list of sequences and a list of barcodes, trim the sequences to the length of the\n",
    "        barcodes\n",
    "        \n",
    "        :param complement: If True, the reverse complement of the barcode will be used\n",
    "        :return: A list of the trimmed sequences.\n",
    "        \"\"\"\n",
    "     \n",
    "        if complement == True:\n",
    "            readed_main = self.reverse_fastq_reader()\n",
    "            readed_barcodes = self.reverse_barcode_reader()\n",
    "\n",
    "        else:\n",
    "            readed_main = self.forward_fastq_reader()\n",
    "            readed_barcodes = self.forward_barcode_reader()\n",
    "\n",
    "        trimmed = []\n",
    "        for seq in readed_main:\n",
    "            for barcode in readed_barcodes:\n",
    "                index = seq.find(barcode)\n",
    "                if not index == -1:\n",
    "                    length = self.length\n",
    "                    trimmed.append(seq[index:index + length])\n",
    "        return trimmed\n",
    "\n",
    "    def fasta_writer(self, complement):\n",
    "        \"\"\"\n",
    "        Write the trimmed sequences to a fasta file\n",
    "        \n",
    "        :param complement: True or False. If True, the reverse complement of the sequence will be\n",
    "        written\n",
    "        \"\"\"\n",
    "     \n",
    "        if complement == True:\n",
    "            trim = self.trimmer(complement=True)\n",
    "            path_file = self.output_dir + self.output_reverse_name\n",
    "            \n",
    "        else:\n",
    "            trim = self.trimmer(complement=False)\n",
    "            path_file = self.output_dir + self.output_forward_name\n",
    "\n",
    "        file = open(path_file, 'w+')\n",
    "        out = '\\n'.join(['>line_' + str(i+1) + \"\\n\" + j for i,j in enumerate(trim)])\n",
    "        file.write(out)\n",
    "        file.close()\n",
    "\n",
    "    def __call__(self):\n",
    "        \"\"\"\n",
    "        The function takes the forward and the reverse complement of the sequence and writes it to a fasta file\n",
    "        \"\"\"\n",
    "    \n",
    "        self.fasta_writer(True)\n",
    "        self.fasta_writer(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = {\n",
    "    'forward_fastq': \"/mnt/c/Users/pc/Downloads/barcod/qc_processed_barcode_1/N70_r1_qc_processed.fastq\",\n",
    "    'reverse_fastq': \"/mnt/c/Users/pc/Downloads/barcod/qc_processed_barcode_1/N70_r2_qc_processed.fastq\",\n",
    "    'barcode_fasta': \"/mnt/c/Users/pc/Downloads/barcod/qc_processed_barcode_1/b14.fasta\",\n",
    "    'output_dir': \"/mnt/c/Users/pc/Downloads/barcod/qc_processed_barcode_1/\",\n",
    "    'output_forward_name': 'trimmed_forward.fasta',\n",
    "    'output_reverse_name': 'trimmed_reverse.fasta',\n",
    "    'length' : 48\n",
    "    }\n",
    "\n",
    "result = Trim(**args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "result()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "trimmonic = 1739877\n",
    "\n",
    "forward_read = 1567482\n",
    "\n",
    "reverse_read = 1502458"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(147, 1570975),\n",
       " (146, 99795),\n",
       " (151, 20185),\n",
       " (145, 12717),\n",
       " (148, 11352),\n",
       " (150, 5719),\n",
       " (143, 2986),\n",
       " (149, 2914),\n",
       " (144, 2574),\n",
       " (54, 1776),\n",
       " (141, 1372),\n",
       " (139, 799),\n",
       " (142, 756),\n",
       " (135, 615),\n",
       " (133, 560),\n",
       " (137, 537),\n",
       " (132, 425),\n",
       " (131, 351),\n",
       " (114, 313),\n",
       " (136, 311),\n",
       " (134, 296),\n",
       " (128, 273),\n",
       " (138, 260),\n",
       " (124, 237),\n",
       " (127, 236)]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter \n",
    "\n",
    "def count_legth_reads(list):\n",
    "    \"\"\"\n",
    "    Given a list of reads, return a list of tuples, where each tuple is the length of the read and the\n",
    "    number of reads of that length\n",
    "    \n",
    "    :param list: the list of reads\n",
    "    :return: A list of tuples, where each tuple contains the length of the read and the number of reads\n",
    "    with that length.\n",
    "    \"\"\"\n",
    "    size = []\n",
    "    for i in list:\n",
    "        size.append(len(i))\n",
    "\n",
    "    counts = Counter(size)\n",
    "\n",
    "    return sorted(counts.items(), key=lambda i:i[1], reverse=True)\n",
    "\n",
    "count_forwad = count_legth_reads(result.forward_fastq_reader())\n",
    "\n",
    "count_forwad[:25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
